### PR TITLE
p2p/simulations: Refactor events + add p2psim

### PR DESCRIPTION
**Pull request by lmars**
This PR refactors events so that they are a single Go type with a `Type` field (making them easier to serialize as a JSON stream), and adds `p2psim` which is a command-line client for the simulation HTTP API.

This now means simulations can be run entirely via the API (i.e. manage the network, watch events, trigger actions via RPC, all mocking / journalling etc.), so we can write dynamic scripts (e.g. bash / Javascript / Ruby) rather than using static Go binaries.

Some things to note:

* in order for node RPC requests to work via the HTTP API (where there could potentially be multiple concurrent clients using conflicting message IDs), I've added an "RPC multiplexer" which multiplexes multiple RPC clients over the single stdin / stdout node RPC connection
* nodes in the network now have a `Name` field which is a human-friendly identifier that can be used with the API (e.g. `p2psim node show net1 node01`)
* I have removed the cytoscape `SimXXX` data structures (the frontend should just consume a stream of `simulations.Event`, see https://github.com/ethersphere/simple-p2p-d3/pull/8)

### CODE REVIEW

File: p2p/simulations/cmd/p2psim/main.go [hunk 1]
```diff
@@ -0,0 +1,425 @@
+// p2psim provides a command-line client for a simulation API.
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/cmd/p2psim/main.go:1]: I thought it best to keep all the simulation stuff under `p2p/simulations` (people won't want to build this binary unless they are running a simulation), I can move if you feel strongly though.

File: p2p/simulations/examples/p2psim.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/examples/p2psim.sh:1]: This script is specific to the example though (it uses the `ping-pong` service).

I plan to fully document all of this code before merging, so there will be `p2p/simulations/doc.go` and `p2p/simulations/README.md` with examples.

File: p2p/simulations/http.go [hunk 3]
```diff
@@ -48,60 +268,57 @@ func NewServer(config *ServerConfig) *Server {
...
-	if req.Header.Get("Accept") == "text/event-stream" {
-		s.streamNetworkEvents(network, w)
-		return
-	}
-
-	s.JSON(w, http.StatusOK, network.Config())
+	s.JSON(w, http.StatusOK, network)
 }
 
 func (s *Server) StartMocker(w http.ResponseWriter, req *http.Request) {
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/http.go:324]: Yes I'm going to clean up the mocking / journal code in a future PR.

File: p2p/simulations/journal.go [hunk 4]
```diff
@@ -97,10 +99,10 @@ func (self *Journal) Close() {
 	close(self.quitc)
 }
 
-func (self *Journal) append(evs ...*event.TypeMuxEvent) {
+func (self *Journal) append(events ...*Event) {
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/journal.go:102]: For now, I will move it in a future PR.

**Review by zelig — APPROVED** [review]

**Pull request merged**